### ISSUE TITLE
Entity type name is culture-sensitive

### ISSUE THREAD

**Issue by mareklinka** [aspnet/EntityFrameworkCore#12617]
I have a solution with the following setup:

HostA (Asp.Net Core 2.1.1)
HostB (Asp.Net Core 2.1.1)
Domain (POCO classes)
DataAccess (EFCore 2.1.1 with DbContext and repositories)
DataAccess.Migrations (storing only the EF migrations)

Using a particular repository that works with a particular entity type works from HostA and fails from HostB. The projects have the same startup and reference the same projects/packages. The exception HostB is throwing:

```
Exception message: Cannot create a DbSet for 'ScheduleRecord' because this type is not included in the model for the context.
Stack trace:
at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.get_EntityType()
at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.get_EntityQueryable()
at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.System.Linq.IQueryable.get_Provider()
at System.Linq.Queryable.Where[TSource](IQueryable`1 source, Expression`1 predicate)
  ... rest of stack truncated for brevity
```

I haven't yet gotten around to trying to construct a minimal repro, but I will get to it when I have the time.

The strange thing is, my context looks like this:

```
private static void ConfigureScheduleItem(ModelBuilder modelBuilder)
{
    var builder = modelBuilder.Entity<ScheduleRecord>();

    builder.HasIndex(b => new { b.TenantCode, b.TaskId }).IsUnique();
    builder.HasIndex(b => b.TaskDate).IsUnique(false);
}
```
Entity:
```
 public class ScheduleRecord : CustomSpecificBase
    {
        public ScheduleType Description { get; set; }

        public int? Duration { get; set; }

        public long ExpertId { get; set; }

        public Expert Expert { get; set; }

        public string Remark { get; set; }

        public DateTime? TaskDate { get; set; }

        public Guid? TaskId { get; set; }
    }
```
And the DB context:
```
public DbSet<ScheduleRecord> ScheduleRecords { get; set; }
```

So the type is definitely part of the context's model. And it works fine in HostA as well. However, from HostB this always throws the error. When I query `context.Model.GetEntityTypes` in HostB, I can also see the type there correctly.

Using both
```
_context.ScheduleRecords
_context.Set<ScheduleRecord>
```
leads to the same results. The problem reproduces for EF 2.1.1 and 2.1.0. This __DOES NOT__ repro with EF 2.0.3. More weirdness, all other entity types work correctly, but this particular one fails. It also fails after renaming the entity.

### Further technical details
EF Core version: 2.1.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 1803
IDE: Visual Studio 2017 15.7.4

**Comment by ivankarpey** [comment]
for my personal situation, I feel like this could be connected to the version of migrations. We have multiple migrations from 2.0 version and now several from 2.1 and maybe 2.1.1 even. So when trying to access newly added entity (from 2.1+) it fails with `cannot create a DbSet for 'Item' because this type is not included in the model for the context.`

**Comment by mareklinka** [comment]
I had the same theory but removing all existing migrations and rescaffolding them did not resolve the problem. Additionally, this entity has been in the model since the start of the project.

**Comment by ivankarpey** [comment]
Was able to overcome this issue using docker container build. But it's definitely something wrong with efcore or idn, maybe some caching or nuget issue

**Comment by divega** [comment]
*EF Triage:* still hoping to get a repro from @mareklinka.

**Comment by mareklinka** [comment]
I will try my best but this is not a small project and I'm afraid removing too much will fail to repro the issue. Unfortunately, I cant share the project as is, being commercial and under an NDA.

Nevertheless, I'll try to get back to you within the next few days.

**Comment by mareklinka** [comment]
I managed to isolate the issue. I'm sorry I couldn't get it into a form of a unit test, but hopefully a minimal executable solution will suffice. If you want me to try and make this into a test case, let me know and I'll take another look.

[TypeNotPartOfModelRepro.zip](https://github.com/aspnet/EntityFrameworkCore/files/2200749/TypeNotPartOfModelRepro.zip)

To run the code, you will need an instance of SQL Server available (any edition). Go to appsettings.json and change the connection string in there to your database server instance. This is the only change you should need to do - upon running the `Host` project (I don't use IISExpress, I run the console host directly for debugging), the application will run

A few observations:
I'm almost 100% sure this is somehow related to how ASP.NET Core localization plays with our DB context (yeah, it sounds crazy to me too).

If you look at `Startup.cs`, you can see a method `SetLocalizationOptions`. Commenting out its body
```
options.DefaultRequestCulture = new RequestCulture(culture: EnUsCulture);
```
will cause the application to work correctly (the Home/Index action will return 0). Keeping the code as is will cause the observed exception. Changing the culture will also have an effect - "sk-SK" will work, but "de-CH" will fail. This seems to be culture specific. The issue also is I can't reproduce this on one of our servers.

The solution is based on latest stable packages for both EFCore and ASP.NET Core.

System info:

Windows language settings: 
![image](https://user-images.githubusercontent.com/7479226/42804825-5e7f9712-89aa-11e8-8de3-9b30e8c71637.png)

```
.NET Core SDK (reflecting any global.json):
 Version:   2.1.300
 Commit:    adab45bf0c

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.300\

Host (useful for support):
  Version: 2.1.0
  Commit:  caa7b7e2ba

.NET Core SDKs installed:
  2.1.4 [C:\Program Files\dotnet\sdk]
  2.1.100 [C:\Program Files\dotnet\sdk]
  2.1.101 [C:\Program Files\dotnet\sdk]
  2.1.102 [C:\Program Files\dotnet\sdk]
  2.1.103 [C:\Program Files\dotnet\sdk]
  2.1.104 [C:\Program Files\dotnet\sdk]
  2.1.200 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.300 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
```

If I can help with anything else, let me know.

**Comment by mareklinka** [comment]
Ok, after a clever tip from a colleague, I think I managed to track down the root cause. It turns out that everything starts working for `en-US` culture when I rename the entity to something not containing "ch". In Slovak locale, "ch" is considered a single character and is in a different place than expected.

This would explain why switching to sk-SK locale works around the problem, I guess. But keeping to a culture where "ch" is just a "c" + "h" leads to issues. I'm not sure why EF has problems with this, but the tests seem reasonable:

1. Renaming the `ScheduleItem` entity so it doesn't contain "ch" solves the issue
2. Renaming previously working entity so it does contain "ch" stops this entity from working

So the question here is why this distinction bothers EF Core. Workaround would be to rename the problematic entities.

**Comment by brunolau** [comment]
Having the very same issue with similar setup (sk-SK vs en-US). I've been scratching my head for a couple of days on thisone so first of all big thanks to Marek for the excellent catch. A fix earlier than 2.2.0 would be handy as I don't really like having to take care of whether I don't accidently use "ch" in the entity name and this might be a sleeping dragon for some others from our region too that yet have to discover it

**Comment by Jupakabra** [comment]
We are running into this too, our web application starts with default US culture, but after that it can be changed by user to many other ones. If other culture has a different ordering rules especially if letter `Y` goes like ` I Y J`. Only entity types which get different ordering fail, because sorted dictionary just cant find them even though their string representation is the same. Hoping we can get the fix in 2.2.0 or earlier :(

**Comment by ajcvickers** [comment]
See also #13318 as well. When fixing this, make sure to check all places where we sort by strings and do so with the invariant culture.

**Comment by TrieBr** [comment]
Great find! I've spent days trying to figure out why my code runs fine locally on windows, but I'm getting the errors you describe on my Ubuntu 18.04 server. As described in #13318, if you build your model snapshot on one machine and run it on a machine with a different locale will fail.

In my case I built my model shapshot on windows 10 with Locale en-US. My Ubuntu server was set to C.UTF-8. I set it to en_US using this sequence of commands:

```
sudo locale-gen en_US
sudo update-locale LANG=en_US
sudo reboot now
```

After running these commands, my the application runs correctly on my ubuntu server.
